design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** ksqlDB currently only supports keys compatible with the `KAFKA` format. This limits the
+           data ksqlDB can work with. Extending the set of key formats ksqlDB supports immediately
+           opens up the use of ksqlDB with previously incompatible datasets.
+           
+## Motivation and background
+
+Data stored in Kafka has a key and a value. These can be serialized using different formats, but are
+generally use a common format. ksqlDB supports multiple _value_ [data formats][1], but requires the 
+key data format to be the `KAFKA`.  
+
+This limitation is particularly problematic for tables. ksqlDB is unable to access changelog topics
+in Kafka that have non-`KAFKA` formatted keys. As the key of the Kafka record is the `PRIMARY KEY` 
+of the table, it is essential that the key can be read if the changelog is to be materialised into a 
+table. When changelog topics have non-`KAFKA` key formats, the limitation precludes ksqlDB as a solution.
+
+The limitation is less damaging for streams. However, it is still the case that the user looses the
+ability to access the data in the Kafka record's key. If this data is not duplicated in the record's
+value, which generally seems to be the case, then the data is not accessible at all. If the data is
+required, then the limitation precludes ksqlDB as a solution.
+
+As well as unsupported _input_ key formats, ksqlDB is equally precluded should a solution require the 
+_output_ to have a non-`KAFKA` key format. ksqlDB is already often used as the glue between 
+disparate systems, even though it is limited to changing the _values_ format and structure.
+Supporting other key formats opens this up to also transforming the key into a different format. 
+
+In some cases users are able to work around this limitation. This may involve changing upstream code,
+or introducing pre-processing, or, in the case of Connect, using SMTs to convert the key format. All
+such solutions tend to increase the complexity of the system, and generally hurt performance.
+
+To open ksqlDB up to new problems spaces and to drive adoption, ksqlDB should support other key formats. 
+
+## What is in scope
+
+ * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+ * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+ * Addition of new server configuration to provide defaults for key and value formats.
+ * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+   the server configuration provides a default.
+ * Support for additional key column data types, where the key format supports it:
+    * `DECIMAL`
+    * `BOOLEAN`
+ * Support of the following key formats:
+    * `DELIMITED`: single key columns as a single string value.
+    * `JSON` / `JSON_SR`: single key column as an anonymous value, i.e. not within a JSON object.
```

Comment:
Why do we have two values? And what is JSON_SR? (I guess this is just a question for my own education).

Reply:
... it's not nice :(

`JSON` format can _read_ a schema from the SR, but doesn't _write_ to the SR.
`JSON_SR` format both _reads_ and _writes_ schemas to SR.

And internally they use different code, so have different behaviour, e.g. when serializing maps (https://github.com/confluentinc/ksql/issues/6049).

I'm not sure of the _why_. @agavra added it, so I'm sure there was good reason.  He can likely comment more.

Reply:
yeah it was a product decision how we wanted to handle it - the difference is that they are actually different serialization formats and are not byte compatible (one adds the magic byte and the schema header, the other is vanilla json). We could've hacked around this and magically removed the header if it exists, but we decided it would be more confusing to the user than just having two formats

drives me bananas too 🍌

Reply:
Not sure if I understand. For `JSON` if the schema is received from the SR, the must still be some "header" encoding the schema ID to fetch the schema? So how can one for both formats _not_ have a "header"?

Reply:
`JSON` does not support schema registry integration

Reply:
>  So how can one for both formats not have a "header"?